docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,317 @@
+    secretRef `json:"secretRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using a label selector.
+    // If this object is nil, no namespaces will be allowed.
+    // If this object is empty {}, clusters can use this identity from any namespace.
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // Selector is a standard Kubernetes LabelSelector.
+    // This is a standard Kubernetes LabelSelector, a label query over a set of resources.
+    // The result of matchLabels and matchExpressions are ANDed.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VSphereIdentityReference struct {
+    // Kind of the identity.
+    // +kubebuilder:validation:Enum=VSphereClusterIdentity,Secret
+    Kind string
+
+    // Name of the identity
+    // +kubebuilder:validation:MinLength=1
+    Name string
+}
+
+type VSphereClusterSpec struct {
+    ...
+    // +optional
+    IdentityRef *VSphereIdentityReference `json:"IdentityRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+* If IdentityRef is specified, and it references a `VSphereClusterIdentity`, the CRD is fetched and used to create a session
+* The secret referenced within `VSphereClusterIdentity` must be in the controller namespace
+* If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session. The secret must be in the same namespace as the `VSphereCluster`
+* The controller will add the VSphereCluster to OwnerReferences on a Secret that contains the account credentials
+* If IdentityRef is not specified, the controller will fallback to using credentials provided in the controller deployment
+* The session will be cached using the existing logic where the key is server + username + datacenter
+* The `IdentityRef` on a `VSphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
+* The controller will fail to delete a `VSphereClusterIdentity` with an error if the account is used by `VSphereClusters`
+
+#### Clusterctl Changes
+
+Today, clusterctl move operates by tracking objectreferences within the same namespace, since we are now proposing to use cluster-scoped resources, we will need to add requisite support to clusterctl's object graph to track cluster-scoped resources that are used by the source cluster, and ensure they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe referenced across namespaces. If a cluster uses a `Secret` for account credentials, the OwnerReference will get set by the controller and the `Secret` will be moved to the target cluster. Any `VSphereClusterIdentity` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.
```

Comment:
> Any `VSphereClusterIdentity` that's required by the resources being moved, will need to be manually created on the target cluster prior to the move operation.

Is this still a requirement? AFAIK there is no need for manually moving objects (except you explicitly wont to do so)

Reply:
agreed

Reply:
thanks will update.